### ISSUE TITLE
Crst Level Violation on rundown event emit

### ISSUE THREAD

**Issue by sywhang** [dotnet/runtime#643]
Started seeing this Crst violation on rundown in debug builds of the runtime:

```
Assert failure(PID 23572 [0x00005c14], Thread: 18100 [0x46b4]): Consistency check failed: Crst Level violation: Can't take level 7 lock CrstEtwTypeLogHash because you already holding level 5 lock CrstSingleUseLock
FAILED: false

CORECLR! CHECK::Trigger + 0x297 (0x00007ffd`d6041ef7)
CORECLR! CrstBase::IsSafeToTake + 0x695 (0x00007ffd`d6143055)
CORECLR! CrstBase::Enter + 0x24B (0x00007ffd`d614241b)
CORECLR! CrstBase::AcquireLock + 0x2A (0x00007ffd`d60cb02a)
CORECLR! CrstBase::CrstHolder::CrstHolder + 0x3A (0x00007ffd`d60c9bda)
CORECLR! ETW::TypeSystemLog::AddTypeToGlobalCacheIfNotExists + 0x1A0 (0x00007ffd`d615e990)
CORECLR! ETW::TypeSystemLog::LookupOrCreateTypeLoggingInfo + 0x719 (0x00007ffd`d6182cd9)
CORECLR! ETW::TypeSystemLog::ShouldLogType + 0x1C7 (0x00007ffd`d6191fa7)
CORECLR! ETW::TypeSystemLog::LogTypeAndParametersIfNecessary + 0x2AD (0x00007ffd`d6180d2d)
CORECLR! ETW::MethodLog::SendMethodDetailsEvent + 0x2D7 (0x00007ffd`d618d1b7)
    File: E:\runtime\src\coreclr\src\vm\crst.cpp Line: 761
```

**Comment by sywhang** [comment]
Updating the thread with some more info I found while investigating this issue... 

The root cause of this is https://github.com/dotnet/coreclr/pull/27515. The new MethodDetails event tries to take the TypeHash lock when it tries to add a type to the cache we have for emitting type data in rundown, but it does that while holding the CodeHeapIterator lock which is a SingleUseLock, so there is this violation. 

As far as I can see, there isn't really a "good" way to fix this in the short run. A potential solution is to emit these events without storing the type in the cache, but that'll be really slow when we're *not* under rundown (this event will get called during assembly unload as well). 

I will come back to this later since this isn't really blocking any product development.

**Comment by janvorli** [comment]
I am now hitting this locally in any attempt to run coreclr tests with a checked or debug build of runtime on OSX and Linux. I have a clean build of the current master. All tests are failing with this assert. I wonder how is it possible that we are not hitting it in the CI.

**Comment by sywhang** [comment]
> I am now hitting this locally in any attempt to run coreclr tests with a checked or debug build of runtime on OSX and Linux. I have a clean build of the current master. All tests are failing with this assert. I wonder how is it possible that we are not hitting it in the CI.

Do you know if you have any eventing turned on? As far as I know there aren’t any tests that turn on the keywords necessary to hit this issue.

**Comment by kunalspathak** [comment]
Same here. I am getting this once in a while when trying to collect profile information using perfcollect on linux arm64 and dotnet-trace on windows arm64.

**Comment by janvorli** [comment]
@sywhang I have looked at the tests/runtest.sh and it enables event logging by default. I've just found that the tests are not failing if I execute them individually using their .sh files. But if I run them all using `./tests/runtest.sh --coreOverlayDir=/home/janvorli/git/runtime/artifacts/tests/coreclr/Linux.x64.Debug/Tests/Core_Root --testNativeBinDir=/home/janvorli/git/runtime/artifacts/obj/coreclr/Linux.x64.Debug/tests --testRootDir=/home/janvorli/git/runtime/artifacts/tests/coreclr/Linux.x64.Debug --copyNativeTestBin Debug` (launched from the src/coreclr), then they fail.

**Comment by CarolEidt** [comment]
I am also seeing this in my test runs on x64/ux. I had 61 instances of this in a test run with no changes.

**Comment by kg** [comment]
When I do test runs on Linux right now from a clean source checkout I get a ton of these errors in various tests. I'm also using runtest.sh with the recommended arguments.

```Consistency check failed: Crst Level violation: Can't take level 7 lock CrstEtwTypeLogHash because you already holding level 5 lock CrstSingleUseLock```

**Comment by CarolEidt** [comment]
I see this on both x64/ux and arm64/ux. Thanks to the hint from @janvorli I added `--disableEventLogging` to my test run. It still reports them as failures at the end of the output, but not in the rolled-up results, making it much easier to figure out where any new failures are.
FWIW, even though this is not "really blocking any product development", it is pretty painful to deal with.

**Comment by echesakovMSFT** [comment]
Carol brought to my attention the question Jan asked:

> All tests are failing with this assert. I wonder how is it possible that we are not hitting it in the CI.

I would like to answer: we never set COMPlus_EnableEventLog in our Helix runs and the CI doesn't use runtest.sh that sets this variable to 1.

I wonder thought if want to enable the event log in our CI (maybe, in a separate test scenario) so we can catch issues like this?

**Comment by sywhang** [comment]
This issue should now be fixed with #36932.
